### PR TITLE
make specialize smarter (bug 5370).

### PR DESCRIPTION
**Pull request by Matafou**
This PR is a proposition to make tactic specialize able to instantiate any subset of its hypothesis, leaving the others in place, with the same names. The limitation it tries to fix is described in bug 5370 (https://coq.inria.fr/bugs/show_bug.cgi?id=5370).

I have two questions:

1) Do we want this in coretactics or is it already too smart to be "core"?
2) Is there a simpler way to program this? In particular can I avoid the manual lift over rels (see line 2942 below)?

### CODE REVIEW

**Comment by herbelin** [comment]
@Matafou: I like this extension a lot but I have a few questions:
- Why not to consider `specialize` on an arbitrary term the same as on an hypothesis, with an additional preliminary `pose proof`? Currently `specialize (fun A => some-term-dependent-on-A) with (A:=B)` is accepted. Why to stop accepting it?
- Why not to make the extension the default behavior (and get rid of the Boolean `e`)? This would virtually break only when `specialize` is used in situations with backtracking where it is supposed to fail, no?
- What about seizing the opportunity of this change to have `specialize` always leaving its result in the context, even when its argument is a term, rather than the heterogeneous current behavior which leaves an hypothesis quantified on the goal when the argument is a term but not when it is an existing hypothesis? Conversely, we could add a `specialize:` (ssreflect style) which would always leave the resulting statement on the goal, even if initially coming from an existing hypothesis. By doing so, we could even have a pretty uniform schema: `destruct`, `induction`, `apply in`, this new `specialize`, `injection` (in structural mode), `rewrite in`, all working in the context, and we could even provide `destruct:`, `induction:`, `apply: in`, `injection:`, `rewrite: in` to tell to add the new result on the goal, ssreflect style.

Also, this change would probably be more welcome as part of a global project than as an incremental change, as discussed at last working group. (But I'm actually precisely continuing working stepwise offline to such global proposal.)

**Comment by Matafou** [comment]
> Why not to consider specialize on an arbitrary term the same as on an hypothesis, with an additional preliminary pose proof? Currently specialize (fun A => some-term-dependent-on-A) with (A:=B) is accepted. Why to stop accepting it?

Good point. I did not see I made a regression here. 

> Why not to make the extension the default behavior (and get rid of the Boolean e)? This would virtually break only when specialize is used in situations with backtracking where it is supposed to fail, no?

If people agree on this I am all for it.

> What about seizing the opportunity of this change to have specialize always leaving its result in the context, even when its argument is a term, rather than the heterogeneous current behavior which leaves an hypothesis quantified on the goal when the argument is a term but not when it is an existing hypothesis?

Good idea. I can modify this PR for that. The would break where specialize behavesl ike generalize, so easy to fix.

> Conversely, we could add a specialize: (ssreflect style) which would always leave the resulting statement on the goal, even if initially coming from an existing hypothesis. By doing so, we could even have a pretty uniform schema: destruct, induction, apply in, this new specialize, injection (in structural mode), rewrite in, all working in the context, and we could even provide destruct:, induction:, apply: in, injection:, rewrite: in to tell to add the new result on the goal, ssreflect style.

I support this (although the colon is already used for goal selector) but I think this is orthogonal to this PR.

> Also, this change would probably be more welcome as part of a global project than as an incremental change, as discussed at last working group. (But I'm actually precisely continuing working stepwise offline to such global proposal.)

The change you propose above about leaving new hyps in the conclusion rather than as new hyps is an interesting global change, but the change I propose is just to make specialize a bit more powerful.

**Comment by herbelin** [comment]
@Matafou: sorry for the digression on colon and a more general model, it was to place my comments within a more general view at tactics, but I agree that this is orthogonal to this PR.

**Comment by Matafou** [comment]
I have a question: this PR actually defines a (not so) new semantics for binding lists. This semantics already exists more or less in eapply. Shouldn't we switch globally to this semantics of binding lists?

**Comment by Matafou** [comment]
@herbelin I am studying your idea of removing the non homogeneous behaviour of specialize when applied on a term instead of a hypothesis. But In the theories there are cases of such uses that cannot be replaced by generalize: when it uses binding lists.

In Sets/Infinite_sets.v near line 163:

specialize Im_inv with (U := U) (V := V) (X := A) (f := f) (y := x).
this cannot be replaced by 
generalize Im_inv with (U := U) (V := V) (X := A) (f := f) (y := x).

because generalize does not support binding list. Of course we could add binding lists to generalize but actually we should add them to *any tactic having terms as arguments*.

I propose that this idea becomes the subject of another PR.

We need to think about it more deeply. I think, like @ejgallego suggest in another PR that binding lists should go in some Ltac syntax for terms instead of being a special argument for tactics with an adhoc treatment everywhere.

**Comment by RalfJung** [comment]
Honestly I feel it is rather strange for `specialize` to accept non-hypotheses as arguments anyway. I feel like it'd make more sense to reject that usage, and tell people to use `pose proof` instead. Essentially, the only difference between `pose proof` and `specialize` is that the latter replaces an existing hypothesis, while the former adds a new one -- correct?
I am not sure how `generalize` fits into this picture.

**Comment by Zimmi48** [comment]
@RalfJung Well actually, `specialize` supports an `as` clause which allows it to behave like `pose proof`. I would personally be in favor of merging these two tactics. I tend to prefer the name `specialize` except when I'm not specializing anything of course: `specialize I as H.` feels weirder than `pose proof I as H.`

Also note that with the `eassert` tactic that @herbelin proposes or the alternative that you are discussing (allowing `assert` to support `_`), one could also do the same with `eassert _ as H by eexact I.` or `assert _ as H by exact I.` A bit more verbose but maybe the intent is actually clearer?

**Comment by RalfJung** [comment]
> I would personally be in favor of merging these two tactics. I tend to prefer the name specialize except when I'm not specializing anything of course: specialize I as H. feels weirder than pose proof I as H.

I guess the term "specializing" is not formally defined. ;)  I don't see any difference between `pose proof I as H` and `pose proof (I x) as H`; in both cases I am giving a term of some type and want that to be recorded in my context.

>  one could also do the same with eassert _ as H by eexact I. or assert _ as H by exact I. or assert (H : _) by exact I. A bit more verbose but maybe the intent is actually clearer?

I don't feel it becomes much clearer if at all -- specialize/pose proof indicate that was are now referring to a previously proven fact that we want to be added to our context, whereas `assert` indicates that we are stating a small lemma and proving it inline. Making the entire goal an evar feels like abusing that feature, but I guess that's a matter of taste. ;)

**Comment by JasonGross** [comment]
> I would personally be in favor of merging these two tactics.

Please don't change `pose proof` to mean ` specialize `; I frequently do `pose proof H as H'` to duplicate `H`.

**Comment by RalfJung** [comment]
> I frequently do pose proof H as H' to duplicate H.

Right, that's why I said above that IMHO both specialize and pose proof make sense, the key difference being that specialize *replaces* a hypothesis instead of adding a new one -- which immediately implies that it can only be applied to things that are already in the context.

**Comment by Zimmi48** [comment]
@JasonGross I never said that one should change the semantics of `pose proof`. I just said that the two already had the same semantics (when you use the `as` clause). E.g. you can do `specialize H as H'.`
But listen: just forget about what I said if it upsets you. It was just a remark in passing.

**Comment by JasonGross** [comment]
@Zimmi48, ah, okay, glad to hear that I had just misinterpreted what you had written.

**Comment by Matafou** [comment]
Merged with @ppedrot econstr.

**Comment by Matafou** [comment]
updated to current trunk.

**Comment by Matafou** [comment]
#248 raises interesting questions about making e-tactics something more systematic but I think this PR can still be merged meanwhile.

File: tactics/tactics.ml
Comment (Zimmi48) [inline tactics/tactics.ml]: typo here: you mean "whatever their place"

Comment (Zimmi48) [inline tactics/tactics.ml]: I also think we say "an hypothesis"

File: tactics/tactics.ml
Comment (Zimmi48) [inline tactics/tactics.ml]: typo: "are supposed to be"

File: tactics/tactics.ml
Comment (Zimmi48) [inline tactics/tactics.ml]: What do you mean here? I can't see any `e` variable.

**Review by Zimmi48 — CHANGES_REQUESTED** [review]
I also think that the extended behavior of this tactic can be evaluated independently from a general plan for e-tactics (and I like this PR).
However, I wouldn't like any new feature / change in behavior to be merged without being documented anymore. Please. (And when I say documented, I mean in the refman, not just in the code, even though documentation in the code is always good).

**Comment by Matafou** [comment]
Good points @Zimmi48 Ill correct all these and document in refman.
and CHANGE too?

**Comment by Zimmi48** [comment]
Yeah, I think it would be good if it were mentioned in CHANGES as well.

**Comment by Matafou** [comment]
There is only a "beyond 8.6" section in CHANGES. This feature will not me merged in 8.7 will it?

**Comment by Zimmi48** [comment]
@Matafou Why not? If it gets merged during the month of May, it will be in 8.6. I can't see any reason for it not to get merged quickly.

**Comment by Matafou** [comment]
In 8.6?
This is not a bug fix, this is an extension of a feature. It should not go in a previously released version of coq.
I thought 8.7 was already in feature freeze.

**Comment by maximedenes** [comment]
> I thought 8.7 was already in feature freeze.

No, it's not. You can find the schedule here: https://github.com/coq/roadmaps/blob/master/text/roadmap-8.7.md

So no problem considering merging this PR for 8.7.

**Comment by Zimmi48** [comment]
@Matafou Sorry I fixed the typo, I meant 8.7 indeed.

**Comment by Matafou** [comment]
Just added tests in test-suite for the new behaviour. Mergeable no I guess. Thanks all for the remarks.

File: test-suite/success/specialize.v
Comment (Zimmi48) [inline test-suite/success/specialize.v]: If this is fixed. Why let it commented?
And if you modify that part, can you clean up the French comments at the same time? I don't even understand what they mean...

File: doc/refman/RefMan-tac.tex
Comment (Zimmi48) [inline doc/refman/RefMan-tac.tex]: It seems to me that this is not based on the latest doc (I made some changes to the doc of `specialize` very recently). Could you rebase on trunk?

**Review by Zimmi48 — COMMENTED** [review]
You didn't add anything to CHANGES finally?
Also, some commits would need squashing.

**Comment by Zimmi48** [comment]
@Matafou did you just push a documentation commit to the wrong branch?

**Comment by Matafou** [comment]
I did push on the wrong branch and then rebased to fix.
Is there still something wrong now?
I will rebase but no conflict is shown.

**Comment by Matafou** [comment]
I rebased, everything seems to be ok. Strangely I had no conflict anuwhere. My modification must have been exactly on lines that you did not touch miraculously.
I squashed too.

File: doc/refman/RefMan-tac.tex
Comment (Zimmi48) [inline doc/refman/RefMan-tac.tex]: I am not satisfied by the result of the automatic merge. I would have put this paragraph before the previous one, because the `as` clause has nothing to do in particular with the first form of `specialize`.

File: doc/refman/RefMan-tac.tex
Comment (Zimmi48) [inline doc/refman/RefMan-tac.tex]: I don't think that this needs to be repeated here.

**Review by Zimmi48 — COMMENTED** [review]
On the question of the squashing I think it would make more sense to squash the typo fixing in the first commit (the one which introduced the typos) instead of the documentation commit (which is unrelated).

File: doc/refman/RefMan-tac.tex
Comment (Matafou) [inline doc/refman/RefMan-tac.tex]: removed.

File: doc/refman/RefMan-tac.tex
Comment (Matafou) [inline doc/refman/RefMan-tac.tex]: moved

File: tactics/tactics.ml
Comment (Zimmi48) [inline tactics/tactics.ml]: Sorry not to have noticed this earlier but why the additional indentation here?

**Review by Zimmi48 — APPROVED** [review]
Thanks for taking time to take my comments into account! This now looks great and I look forward to this PR being merged.

**Comment by herbelin** [comment]
BTW, this PR fixes a bug of `specialize` which would help #637.

File: tactics/tactics.ml
Comment (Matafou) [inline tactics/tactics.ml]: dunno.

Comment (Zimmi48) [inline tactics/tactics.ml]: Well, since you are the one who introduced this indentation, may I suggest that you fix it before this is merged? This will avoid an irregular style in this file. It should take just one commit, to be squashed with the first one in this PR.

**Comment by herbelin** [comment]
Update: In fact, one of the examples I'm looking at (derived from an example in CoRN) fails in this branch:
```coq
Goal (forall x y, x=0 -> y=0 -> True) -> True.
intros H.
specialize (fun z => H 0 z eq_refl) with (z:=0).
(* anomaly *)
```
while before a similar example was giving non sense:
```coq
Goal (forall x y, x=0 -> y=0 -> True) -> True.
intros H.
specialize (fun z => H 0 z) with (1:=eq_refl).
(* "H : forall y : nat, 0 = 0 -> y = 0 -> True" while "H : forall y : nat, y = 0 -> True" is expected *)
```
I don't know what to think though. The example does not look so interesting to have. Maybe your initial idea that `with` should be restricted to the case of a reference is a better direction to go.

**Comment by Zimmi48** [comment]
Note that the use of specialize with a non-ident is undocumented. Is it very much used? Should it get documented? Does it have test cases?

**Comment by herbelin** [comment]
A grep of "specialize (fun" in coq-contribs gives 3 occurrences (1 in intuitionistic-nuprl, 1 in corn, 1 in cfgv). There are 2 in test-suite (4221.v and 4456.v, both from fiat).

Another worry is the consistency with "constr with bindings" in general. There were actually a discussion with Ralf Jung on this subject on coq-club, with somehow the conclusion that "with" is not so much intuitive for a constr different from a reference. But if we want to revise this, we shall stumble again on the question of versioning tactics.

**Comment by RalfJung** [comment]
@herbelin are you referring to <https://coq.inria.fr/bugs/show_bug.cgi?id=4536>? What makes that one worse is that `lemma1` *looks* like is just a reference, and only "thanks" to maximally inserted implicit arguments, it is not.

**Comment by Matafou** [comment]
Indeed this case is hard to deal with. Note however that the (fun z => ..) trick is precisely what this PR tries to do automatically and discharge from the user. So maybe we can make it forbidden? It introduces a hopefully rare incompatibility but workaround should simplify the code (use incomplete with_binding instead of fun =>...)?

I will try to fix this next week, either by failing gracefully or by supporting it correctly.

P.

**Comment by JasonGross** [comment]
> A grep of "specialize (fun" 

I use this form a lot in fiat-crypto and fiat-parsers (though only with hypotheses at the head under the lambda)

**Comment by Zimmi48** [comment]
> Note however that the (fun z => ..) trick is precisely what this PR tries to do automatically and discharge from the user. So maybe we can make it forbidden? It introduces a hopefully rare incompatibility but workaround should simplify the code (use incomplete with_binding instead of fun =>...)?

@JasonGross is @Matafou's assumption correct? If yes, I would favor this solution but that would require a compatibility setting.

**Comment by JasonGross** [comment]
Frequently yes, but here are two important use cases this PR doesn't cover:
```
match goal with
| [ H : forall x : _ * _, _ |- _ ]
  => specialize (fun a b => H (a, b))
| [ H : forall x f, ?P x = f -> _ , H' : forall z, ?P (@?p z) = @?Q z |- _ ]
  => specialize (fun z => H (p z) (Q z) (H' z))
end
```

**Comment by Matafou** [comment]
@JasonGross this should work. It is mixing (fun x => ...) and with-bindings that makes the bug appears.

**Comment by JasonGross** [comment]
@Matafou Great!  I personally never use with-bindings with specialize (though perhaps I'll start after this PR lands), so I don't mix them with `fun`

**Comment by Matafou** [comment]
I think I fixed it. i had to put a let in to put the two sources of arguments apart.

**Comment by Matafou** [comment]
I think all remarks have been implemented for this PR. It seems to also solve some previous issues with specialize, namely the mixing of using a term starting with an abstraction and with-binding.

A nice example of a PR that gained A LOT from being reviewed. Thanks all.

File: tactics/tactics.ml
Comment (ppedrot) [inline tactics/tactics.ml]: Why use a dummy identifier when you can use an anonymous let-in?

File: tactics/tactics.ml
Comment (ppedrot) [inline tactics/tactics.ml]: Please don't use unsafe typing functions. Use `pf_type_of` or `pf_get_type_of` depending whether you assume the term to be well-typed in the context.

File: tactics/tactics.ml
Comment (ppedrot) [inline tactics/tactics.ml]: Same thing about unsafe typing here.

**Review by ppedrot — CHANGES_REQUESTED** [review]
As a general remark, this tactic uses legacy meta API and will be rewritten at some point. We will merge it for 8.7 but reimplement it at some point afterwards.

**Comment by maximedenes** [comment]
Summary of WG discussion: we merge this after the code has been improved a bit (see @ppedrot's remarks). It is a pity to introduce new code based on the old engine, but rewriting `specialize` will take some time.

**Comment by Matafou** [comment]
OK. Will do it this week.

I volonteer to port it to the new API after 8.7 (with some help maybe) is
out.
P.

2017-05-24 21:16 GMT+02:00 Maxime Dénès <notifications@github.com>:

**Comment by Zimmi48** [comment]
@Matafou Don't forget that the freeze is Thursday (and tests need to re-run before the merge). I wouldn't like this missing the 8.7 window for no real reason.

**Comment by Matafou** [comment]
Ok will try tomorrow (Wednesday).
P
Le mar. 30 mai 2017 à 11:33, Théo Zimmermann <notifications@github.com> a
écrit :

**Comment by Matafou** [comment]
mmh I rebased on a an old trunk and got allt hose strange commits above.

I rebased on the right trunk and everything seems fine.

The remarks of @ppedrot have been implemented.

**Comment by Matafou** [comment]
@ppedrot can you have a look at line 3017 (commit 6378c87 above) and confirm that I should indeed pass clause.evd as the initial sigma?
The real question is: is clause.evd really

**Comment by Matafou** [comment]
Sorry for the mess with the PR interface. I am having a hard morning :)

**Comment by ppedrot** [comment]
Looks OK to me. If I could add a last nitpick, it would be even better to factorize all the uses of `clause.evd` by adding a `let sigma = clause.evd in` somewhere (e.g. after the call to `clenv_unify_meta_types`). If you don't have time, that's fine, the PR can be merged as is.

**Comment by Matafou** [comment]
Done.

This time I think we can merge (after tests).

**Comment by ejgallego** [comment]
@ppedrot I understand you can remove your "needs change" flag now, right?

**Comment by Zimmi48** [comment]
The way to do so is by approving the PR explicitly (not really a flag).

**Review by ppedrot — APPROVED** [review]

**Comment by maximedenes** [comment]
I took care of the merge, it is being tested here:
https://travis-ci.org/maximedenes/coq/builds/238357217

**Comment by Matafou** [comment]
Cool. Can I close the PR then?

**Comment by maximedenes** [comment]
> Cool. Can I close the PR then?

No, please leave it open so that I don't forget. I'll close it when doing the final merge.

Thanks!

**Pull request merged**
